src/main/java/com/aws/greengrass/componentmanager/DependencyResolver.java

```diff
@@ -221,11 +222,18 @@ void checkNonExplicitNucleusUpdate(List<String> targetComponents,
             ComponentMetadata resolvedVersion = componentResolver.resolve(componentToResolve, versionConstraints);
             logger.atDebug().kv("resolvedVersion", resolvedVersion).log("Resolved component");
             resolvedComponents.put(componentToResolve, resolvedVersion.getComponentIdentifier());
-            resolvedVersion.getDependencies().forEach((k, v) -> {
-                componentNameToVersionConstraints.putIfAbsent(k, new HashMap<>());
-                componentNameToVersionConstraints.get(k).put(componentToResolve, Requirement.buildNPM(v));
-                componentsToResolve.add(k);
-            });
+
+            for (Map.Entry<String, String> dependency : resolvedVersion.getDependencies().entrySet()) {
+                // A circular dependency is present if the dependency is already resolved.
```

Comment:
Is this really the case? can the dependency not be previously resolved if it was a dependency of some other component, i.e. in an independent branch of the dependency tree?

Reply:
resolvedComponents is only tracking the dependencies of one root component. It does not track resolved components across multiple . But you are right, in the same dependency tree if the component was resolved because it was dependency of some other component it will fail.


Below example which I used to test works, but if A depends on C1 then it fails. Let me fix that.
               group1
                    |(1.0.0)
                   A
       (1.0.0)/   \(>1.0)
              B1     B2
      (1.0.0)\    /(1.0.0)
                 \  /
                  C1